File: JuliaCode/randomTree.jl
function randomTree(m, mixedDirection, outflow_pattern, inflow_pattern)
    p = m+1
    M = shiftopZeros(ShiftOperator,p,m)
    M_unidirected = shiftopZeros(ShiftOperator,p,m)
    for i = 1:m
        # Add edge
        base_node = rand(1:i)
        new_node = i + 1

        random_direction = rand(0:1)

        if !mixedDirection
            random_direction = 1
        end

        if random_direction == 1
            M[base_node,i] = outflow_pattern
            M[new_node,i] = inflow_pattern
        else
            M[new_node,i] = outflow_pattern
            M[base_node,i] = inflow_pattern
        end

        if mixedDirection
            M_unidirected[base_node,i] = outflow_pattern
            M_unidirected[new_node,i] = inflow_pattern
        end
    end
    M = M[end:-1:1, end:-1:1]
    M_unidirected = M_unidirected[end:-1:1, end:-1,1]

    # Reorder edges and nodes to make it more logical
    source_node = p
    if mixedDirection
        improved_rev_orderN, improved_rev_orderE = findRevOrder(M_unidirected, source_node, outflow_pattern, inflow_pattern)
    else
        improved_rev_orderN, improved_rev_orderE = findRevOrder(M, source_node, outflow_pattern, inflow_pattern)
    end
    improved_orderN = improved_rev_orderN[end:-1:1]
    push!(improved_orderN, p)
    improved_orderE = improved_rev_orderE[end:-1:1]
    M = M[:, improved_orderE]
    M = M[improved_orderN, :]
    return M
end

function findRevOrder(M, base_node, outflow_pattern, inflow_pattern)
    # Find connected edges
    edges = findPattern(M[base_node, :], outflow_pattern)

    # Add edges
    orderE = Int[]
    orderN = Int[]
    for e = edges
        # Find new base node
        outflow_node = findPattern(M[:,e], inflow_pattern)
        push!(orderN, outflow_node[1])
        push!(orderE, e)
    end

    # Recursive part
    for e = edges
        # Find new base node
        outflow_node = findPattern(M[:,e], inflow_pattern)
         
        # Find order from new base node
        subgraph_orderN, subgraph_orderE = findRevOrder(M, outflow_node, outflow_pattern, inflow_pattern)

        # Append order
        for ind = subgraph_orderN
            push!(orderN, ind)
        end
        for ind = subgraph_orderE
            push!(orderE, ind)
        end
    end
    return orderN, orderE
end

function findPattern(V, op)
    # Returns all entry indices in M that are equal to op
    n = length(V)
    is = Int[]
    for i = 1:n
        vi = V[i]
        if vi == op
            push!(is, i)
        end
    end
    return is
end

File: MatlabCode/randomTree.m
function M = randomTree(m, mixedDirection, outflow_pattern, inflow_pattern)
%RANDOMTREE creates a matrix of shift operators corresponding to a random
%tree graph
%
%RANDOMTREE(m, mixedDirection, outflow_pattern, inflow_pattern)
%   m - The number of edges in the graph
%   mixedDirection - If true (false) the direction of the edges will be
%                       random (deterministic)
%   outflow_pattern - The element in the matrix to represent outflow
%   inflow_pattern - The element in the matrix to represent inflow
    p = m + 1;
    M = shiftOperator.zeros(p,m);
    M_unidirect = shiftOperator.zeros(p,m);
    for i = 1:m
        % Add edge
        base_node = randi([1,i],1); % Random base node
        new_node = i + 1;
        
        random_direction = randi([0,1],1);
        if ~mixedDirection
            random_direction = 1;
        end
        
        if random_direction
            M(base_node, i) = outflow_pattern;
            M(new_node, i) = inflow_pattern;
        else
            M(new_node, i) = outflow_pattern;
            M(base_node, i) = inflow_pattern;
        end
        
        if mixedDirection
            M_unidirect(base_node, i) = outflow_pattern;
            M_unidirect(new_node, i) = inflow_pattern;
        end
    end
    M = M(end:-1:1,end:-1:1);
    M_unidirect = M_unidirect(end:-1:1,end:-1:1);

    % Reorder edges and nodes to make it more logical
    source_node = p;
    if mixedDirection
        [improved_rev_orderV, improved_rev_orderE] = findRevOrder(M_unidirect, source_node, outflow_pattern, inflow_pattern);
    else
        [improved_rev_orderV, improved_rev_orderE] = findRevOrder(M, source_node, outflow_pattern, inflow_pattern);
    end
    improved_orderE = improved_rev_orderE(end:-1:1);
    improved_orderV = [improved_rev_orderV(end:-1:1), p];
    M = M(:, improved_orderE);
    M = M(improved_orderV,:);
end

function [orderV, orderE] = findRevOrder(M, base_node, outflow_pattern, inflow_pattern)
    % Find connected edges
    [~, edges] = findPattern(M(base_node,:), outflow_pattern);

    % Add edges
    orderE = edges;
    orderV = [];
    % Add nodes
    for e = edges
        % Find new base node
        outflow_node = findPattern(M(:,e), inflow_pattern);
        orderV = [orderV, outflow_node];
    end

    % Recursive part
    for e = edges
        % Find new base node
        outflow_node = findPattern(M(:,e), inflow_pattern);
        
        % find order from there
        [subgraph_orderV, subgraph_orderE] = findRevOrder(M, outflow_node, outflow_pattern, inflow_pattern);
        
        % Append order
        orderE = [orderE, subgraph_orderE];
        orderV = [orderV, subgraph_orderV];
    end
end

function [is,js] = findPattern(M, op)
    % returns all entry indecies in M that are equal to op
    [n,m] = size(M);
    is = [];
    js = [];
    for i = 1:n
        for j = 1:m
            mij = M(i,j);
            if mij == op
                is(end+1) = i;
                js(end+1) = j;
            end
        end
    end
end
